ros/src/twist_controller/dbw_node.py

```diff
@@ -56,49 +69,135 @@ def __init__(self):
-
-        # TODO: Subscribe to all the topics you need to
-        rospy.Subscriber('/vehicle/dbw_enabled', Bool, self.dbw_enabled_cb)
+        self.controller = Controller(vehicle_mass,
+                                     fuel_capacity,
+                                     brake_deadband,
+                                     decel_limit,
+                                     accel_limit,
+                                     wheel_radius,
+                                     wheel_base,
+                                     steer_ratio,
+                                     max_lat_accel,
+                                     max_steer_angle)
+
+        rospy.Subscriber('/current_velocity',
+                         TwistStamped,
+                         self.current_velocity_cb)
+
+        rospy.Subscriber('/twist_cmd',
+                         TwistStamped,
+                         self.twist_cmd_cb)
+
+        rospy.Subscriber('/vehicle/dbw_enabled',
+                         Bool,
+                         self.dbw_enabled_cb)
+
+        rospy.Subscriber('/current_pose',
+                         PoseStamped,
+                         self.current_pose_cb)
+
+        rospy.Subscriber('/final_waypoints',
+                         Lane,
+                         self.waypoints_cb)
+
+        self.activated = False
+        self.current_velocity = None
+        self.proposed_velocities = None
+        self.current_pose = None
+        self.waypoints = None
 
         self.loop()
 
     def loop(self):
-        rate = rospy.Rate(50)  # 50Hz
+        rate = rospy.Rate(10)  # originally 50Hz
         while not rospy.is_shutdown():
-            # TODO: Get predicted throttle, brake, and steering using `twist_controller`
-            # You should only publish the control commands if dbw is enabled
-            throttle, brake, steering = (1.0, 0.0, 0.0)
-            # throttle, brake, steering = self.controller.control(<proposed linear velocity>,
-            #                                                     <proposed angular velocity>,
-            #                                                     <current linear velocity>,
-            #                                                     <dbw status>,
-            #                                                     <any other argument you need>)
-            if self.dbw_enabled:
-                self.publish(throttle, brake, steering)
+
+            if self._valid_state():
+                # rospy.logwarn('p.v. linear : %f, p.v. angular: %f, c.v. : %f',
+                #               self.proposed_velocities.twist.linear.x,
+                #               self.proposed_velocities.twist.angular.z,
+                #               self.current_velocity.twist.linear.x)
+
+                with self.lock:
+                    cte = compute_cte(self.waypoints, self.current_pose)
+
+                throttle, brake, steer = self.controller.control(self.activated,
```

Comment:
I think it would be better to assign self.activated to a variable and hand it over to self.controller.control and in line 129.
Just to ensure consistency and avoid corner cases where there is switch while being in self.controller.control()

Reply:
yes.. good call.. will do that.